pruner/pruner.go

```diff
@@ -0,0 +1,359 @@
+package pruner
+
+import (
+	"sync"
+	"time"
+
+	quamina "quamina/lib"
+)
+
+// Stats reports basic counts to aid in deciding when to Rebuild.
+type Stats struct {
+	// Some of these values are ints instead of uints because Go
+	// likes to print uints in hex, and I'd like to see some
+	// temporary logging output in decimal.  ToDo: back to uints?
+
+	// Live the count of total live patterns.
+	Live int
+
+	// Added is the count of total patterns added.
+	Added int
+
+	// Deleted is the count of the patterns removed.
+	Deleted int
+
+	// Emitted is the count to total patterns found since the last
+	// rebuild.
+	Emitted int64
+
+	// Filtered is the count of pattners that have been removed
+	// from MatchFor results (since the last rebuild) because
+	// their patterns had been removed.
+	Filtered int64
+
+	// LastRebuilt is the time the last rebuild started.
+	LastRebuilt time.Time
+
+	// RebuildDuration is the duration of the last rebuild.
+	RebuildDuration time.Duration
+
+	// RebuildPurged is the count of patterns removed during
+	// rebuild.
+	RebuildPurged int
+}
+
+// Matcher provides DeletePattern on top of quamina.Matcher.
+//
+// Matcher maintains the set of live patterns, and it will rebuild the
+// underlying matcher synchronously (currently) periodically sduring
+// standard operations (AddPattern, DeletePattern,
+// MatchesForJSONEvent).
+//
+// Roughly speaking, the current rebuild policy automatically rebuilds
+// the index when the ratio of filtered patterns to emitted patterns
+// exceeds 0.2 (and if there's been some traffic).
+//
+// An application can call Rebuild to force a rebuild at any any.  See
+// Stats() to obtain some useful statistics about the matcher.
+//
+// Eventually automatically-invoked rebuild policies might be
+// pluggable.
+type Matcher struct {
+	// Matcher is the underlying matcher that does the hard work.
+	//
+	// Maybe Matcher should maybe not be embedded or public.
+	Matcher *quamina.CoreMatcher
+
+	// live is live set of patterns.
+	live LivePatternsState
+
+	stats Stats
+
+	// rebuildTrigger, if not nil, determines when a mutation
+	// triggers a rebuild.
+	//
+	// If nil, no automatic Rebuild is ever triggered.
+	rebuildTrigger rebuildTrigger
+
+	// lock protectes the pointer the underlying Matcher as well as stats.
+	//
+	// The Matcher pointer is updated after a successful Rebuild.
+	// Stats are updated by Add, Delete, and Rebuild.
+	lock sync.RWMutex
+}
+
+var defaultRebuildTrigger = newTooMuchFiltering(0.2, 1000)
```

Comment:
That's the filtered/emitted min ratio and min number of total X's emitted.

What should these constants be?

Reply:
*snicker* what makes you think I have a better idea. Might as well run with these.